requestsignature/requestsignature.go

```diff
@@ -0,0 +1,328 @@
+	if authorizationHeaderValue == "" {
+		return "", errors.New("authorization header missing")
+	}
+	if !bearerRegex.MatchString(authorizationHeaderValue) {
+		return "", errors.New(fmt.Sprintf("found authorization header is not a valid Bearer token. Got %v", authorizationHeaderValue))
+	}
+	authorizationHeaderValue = strings.TrimPrefix(authorizationHeaderValue, "Bearer ")
+	return authorizationHeaderValue, nil
+}
+
+func (signer *requestSigner) isTimestampValid(req *http.Request) error {
+	timestampHeaderValue, err := time.Parse(time.RFC3339, req.Header.Get("x-dv-signature-timestamp"))
+	if err != nil {
+		return err
+	}
+
+	timeNow := signer.now().UTC()
+
+	timeBeforTimestamp := timeNow.Add(-timeDiff)
+	timeAfterTimestamp := timeNow.Add(timeDiff)
+
+	if !(timestampHeaderValue.After(timeBeforTimestamp) && timestampHeaderValue.Before(timeAfterTimestamp)) {
+		return errors.New(fmt.Sprintf("request is timed out: timestamp from request: %v, current time: %v", timestampHeaderValue.Format(time.RFC3339), timeNow.Format(time.RFC3339)))
+	}
+
+	return nil
+}
+
+func (signer *requestSigner) getHexHashForNormalizedHeaders(req *http.Request) (hex string, err error) {
+	if req.Body == nil {
+		return "", errors.New("payload missing")
+	}
+
+	body, err := signer.getBodyFromRequest(req)
+	if err != nil {
+		return "", err
+	}
+
+	signedHeaders := strings.Split(req.Header.Get(signatureHeaderKey), ",")
+	sort.Strings(signedHeaders)
+
+	normalizedRequest := signer.getNormalizedRequestWithHeaderAndBody(req, signedHeaders, body)
+
+	strNormalizedRequest := strings.Join(normalizedRequest, "\n")
+	log.Printf("normalized request: %#v", strNormalizedRequest)
+
+	hashNormalizedRequest := sha256.Sum256([]byte(strNormalizedRequest))
+
+	return strings.ToLower(fmt.Sprintf("%x", hashNormalizedRequest)), nil
+}
+
+func (signer *requestSigner) getNormalizedRequestWithHeaderAndBody(req *http.Request, signedHeaders []string, body []byte) []string {
+	normalizedHeaders := []string{}
+
+	for _, name := range signedHeaders {
+		headerValue := req.Header.Get(name)
+		normalizedHeaders = append(normalizedHeaders, fmt.Sprintf("%v:%v", strings.ToLower(name), strings.TrimSpace(headerValue)))
+	}
+
+	normalizedRequest := []string{}
+	normalizedRequest = append(normalizedRequest, req.Method)
+	normalizedRequest = append(normalizedRequest, req.URL.Path)
+	normalizedRequest = append(normalizedRequest, req.URL.RawQuery)
+	normalizedRequest = append(normalizedRequest, fmt.Sprintf("%v\n", strings.Join(normalizedHeaders, "\n")))
+	normalizedRequest = append(normalizedRequest, signer.getHexHashedPayload(body))
+
+	return normalizedRequest
+}
+
+func (signer *requestSigner) getBodyFromRequest(req *http.Request) ([]byte, error) {
+	if req.GetBody != nil {
+		log.Print("get copy of request body")
+
+		var bodyReader io.Reader
+		bodyReader, err := req.GetBody()
+		if err != nil {
+			return nil,  err
+		}
+
+		body, err := ioutil.ReadAll(bodyReader)
```

Comment:
At this point you read the body of the request. The body will not be visible for other HTTP-Handler. You can simply add the read body as a NopCloser back to the Request. 

req.Body = ioutil.NopCloser(bytes.NewReader(body))